### DISCUSSION TITLE
Allow a global installation folder

### DISCUSSION THREAD

**Discussion by DavinChurch** [aplteam/Tatin#55] [Ideas]
I have toolkits that I make use of in many applications. I don't want to have to install duplicates of each of these toolkits in every application directory on my system. In fact, some of my small applications don't even reside in separate directories, and I may have hundreds of them.

This could also be an issue when I have sub-applications that require different sets of packages for support of the different layers. I don't want to have to load all the packages into the top-level application, nor do I want to have to load duplicates of everything throughout the application structure.

What I'd like to be able to do for such a situation is be able to install all the toolkits I use in a global directory (such as C:\Packages). Then I can use LoadDependencies in each application to bring in the tools that I need for that application.

Actually, I can force Tatin to do most of that for me, even though it's not part of the basic design. But where it becomes more of a problem is that when I LoadDependencies I get **ALL** the packages loaded from the global store, even those I don't want. So to improve this situation, I'd like a way to load only selected packages into a given application. I think this could be accomplished by simply changing or adding to the parameters for LoadDependencies, without any other modifications to Tatin. If this were made optional, then it could be used both in the way I describe here as well as in the way it was originally designed, without compromise.

One possible method for accomplishing this is to allow the package-containing directory name to also include the (full or partial) name of a particular package, to indicate that only that package is to be loaded. Alternatively, the selected package(s) could be optionally added to the end of command line arguments.

This seems to be a relatively simple request that doesn't impact existing functionality but provides more flexibility for users that wish to use Tatin in slightly different ways.

**Comment by aplteam** [comment]
This goes against the grain of Tatin. 

@morten? @Gil?

**Comment by DavinChurch** [comment]
I don't see how this is detrimental to any designed use of Tatin. It's simply an extension to allow the user a choice in how to use it. It doesn't prevent the original method of doing things at all. A reasonable (and non-harmful) choice is good, is it not?

**Comment by DavinChurch** [comment]
Perhaps if I explained my alternative if I don't have an extra optional parameter in Tatin...

Say I've installed 20 packages (in a common area, which is possible for me to do), but I only need one of them in a simple new workspace I'm starting to build. For the simplest case, I LoadDependencies to load all 20 installed packages and then go through and erase 19 of the namespaces from my workspace along with their direct references. This isn't too horrible if I have to do it only once.

However, if my desired package has one or more dependent packages, then I have to go through and figure out the appropriate dependencies manually and erase only the ones that aren't needed. This is one of the primary jobs of Tatin and I'd have to be doing it completely manually. And if I ever have to rebuild the workspace or reinstall new versions then I have to start over and do the entire procedure again from scratch. And again the next time, etc. And that's just for one simple workspace.

So it seems to me a trivial improvement to allow extra argument(s) on the command line to allow me to just load the packages that I want to use (along with their dependencies). It solves all my problems with the above (and similar) scenarios, offers flexible options to all users, changes nothing to anyone using Tatin as originally designed, changes nothing in the overall scope or design or purpose of Tatin, and requires rather trivial changes to the Tatin codebase. I can't see any disadvantages to allowing this option and it can provide considerable functionality and ease-of-use in certain situations, extending Tatin's usefulness overall.

**Comment by aplteam** [comment]
@Morten: if you do not object I am going to implement this as an experimental feature.

**Reply by e9gille** [reply]
I think you are asking the wrong Morten. @mkromberg

**Comment by e9gille** [comment]
I see I managed to post my message in the wrong discussion thread (https://github.com/aplteam/Tatin/discussions/57#discussioncomment-1419310).

Is the reason for this to save space?

The current design is based on each folder specifying its requirements, so there are implications in using a single folder for all applications and then dynamically filtering what to be loaded. This may be exactly what is desired here, but I want to be sure we are solving the right problem.

**Reply by DavinChurch** [reply]
I agree with your design considerations, but that's only one way to make use of a package management system. There are many possible needs for a system like Tatin, and it should be flexible enough to cover more than one design if it can be easily and compatibly extended.

I've tried to mention a few different reasons for allowing alternate organizations, and one of them has to do with size (both space and organization size). Just for an example from my own needs, I'm coming from an APL+Win environment which I've been using on the PC since 1981. I have **8,518** APL+Win workspaces currently on my hard drive! Almost every one uses one or more of my toolkits, and almost none of them have their own separate directory structure (they reside as single workspaces in common directories). So, how many times will I have to install my toolkits on disk to make them accessible to all the different workspaces? And how many artificial directory structures will I have to create to install them to?

I may represent an extreme example of the population out there, but there are bound to be quite a number of APLers that want a simple, shared workspace organization but could still use a nice tool like Tatin to provide them pre-built toolkits to use, don't you think? Isn't that reason enough to allow a little extra flexibility in the design?

**Reply by e9gille** [reply]
Just to be clear, I'm trying to lay out your requirements clearly to see how best to resolve the issues. Yes, I think it is great if there is flexibility in the design, I just can't imagine all the different ways it might be used, so that's what I'll try to wrap my head around now :)

In your case, you have thousands of projects/workspaces using common tools. You state that they might require different versions of the same tool (some using newer features, others trying to avoid newer versions due to bugs introduced). This suggests you don't simply want to use the latest version of a tool across the board. You ask for a way to specify which packages to load into a given workspace dynamically (potentially a specific version even). So for each workspace you need a declaration of dependencies anyway. Why not keep this in a separate file?

During the design of Tatin I argued the case for loading packages dynamically (without having to install them in a folder first). My reason was that I may want to experiment with a package, or use it during a session as a one off exercise to get something done. Tatin allows me to do that, but perhaps not in the most efficient way right now (never optimise/optimise later mantra).

I can see how in your situation you might prefer to have each one of your workspaces call on Tatin to load a set of dependencies dynamically. In theory it doesn't matter how Tatin accomplishes that, but we can imagine a global cache somewhere on your disk (say C:\Packages for example :).

So, how does that differ from your suggestion? Well, the cache wouldn't contain a dependency declaration or build list of any sort. Also, every time your application runs, it may or may not pick the same version of packages depending on how specific you are with the dependency versions to install (say Group-Tool-1.0.0 vs Group-Tool-1). In short, this is probably what you are looking for, but please let me know what I've overlooked.

**Reply by DavinChurch** [reply]
Thank you for your careful considerations. Usually (in my case) I don't have to worry too much about using multiple versions simultaneously, as that occurs pretty rarely for me. But I do sometimes need to be using an older static version. Other people may have different needs along these lines and it would be a shame to not be able to support them as well.

It might help if I explain that I seldom dynamically refresh my tool inclusions like that - they are almost always saved static.ally - though I know other people that do that regularly. You see, I )SAVE my WSes with the version of the tools that I want to use. Later, if one of my toolkits has been updated (which doesn't happen often), I will usually explicitly tell it to update the tool in my saved WS when I happen to be modifying that program anyway, while making sure that everything still works as expected. A few programs I don't update the static version (for various reasons) unless I have a particular need for a new feature and a lot of re-testing is required. New WSes (or new inclusions) will always start with the latest tool version.

Likewise, in my typical situation, I don't really need to track installed-vs-available versions for each package. I'll either leave them as they are or decide to refresh them manually. So I really don't need much tracking most of the time. However, with Tatin available I expect that may change somewhat when using other people's toolkits on a regular basis, and I may need to reevaluate some of my procedures.

Another thing is that my own tool packages almost never have dependencies amongst themselves, so tracking cross-dependency information for each WS is not normally necessary. Once a tool is "linked" to one of my WSes it's effectively complete and operational. That's not to say that I don't make use of a number of tool packages per WS at one time - I usually do.  It's just that the main WS needs them all but they don't need one another.

BTW, I liked your idea of loading packages directly into WSes without installing them on disk first, but Kai tells me that a disk installation is important for (at least) external resources. I'm not used to working with such resources, but I can see that some packages may have a need for such a facility. In most cases I would use it to initially do a semi-permanent loading into my saved WS without having a separate disk copy at all, and update it only on command when I wanted to.

Does that help clarify my particular situation any? Of course I may be using the "standard" way of doing things more with Tatin and/or when building large applications that deserve their own working area. It's just that I haven't needed that sort of thing much in the past.

**Reply by e9gille** [reply]
Well in that case I think we agree, you shouldn't need to install packages in a folder at all. The day this approach causes you issues, please report and describe the scenario and we'll see if a change is required to Tatin.

In any case, @aplteam, I think we might want to open the discussion again on caching. If Tatin used a global cache (in a user configurable location ?), then:
1. dynamic dependency loading would be faster.
2. even with packages that use more advanced features such as external assets, the workspace could be saved and would work between sessions.
3. You could work with packages even if temporarily offline (when using versions already cached).

**Reply by mkromberg** [reply]
Sounds extremely desirable.

**Reply by DavinChurch** [reply]
Thank you. I think that simply loading uninstalled packages would work well for my traditional situation. If I need something with external resources then I'll just have to provide an install area in which to store it. Large or disk-based projects can be given their own installation folders as designed.

**Reply by DavinChurch** [reply]
When you're discussing caching, I can see two possible ways to work ... one is disk caching with a common installation area like I originally requested, and another I think would be handy to consider would be session caching.

For instance (in my case), if I installed all my frequently-used packages into ⎕SE (and perhaps saved them there or set up an auto-install upon starting) instead of saving in a disk directory or loading them in each WS, then I wouldn't need to )SAVE a separate copy in every WS. I could instead just reference the shared ⎕SE namespaces directly from each application as needed. Of course, for anything I would be distributing off my machine I'd be sure to load them directly into the workspace (or install them in the app directory), but I have so many WSes that I just use for myself that I think that ⎕SE caching would be quite beneficial personally while still allowing Tatin to load/update/manage them as dynamically as needed. Of course, such a concept would be a significant enhancement to be considered for a later version of Tatin.

Think that might be worth a thought?

**Reply by e9gille** [reply]
I don't like the idea that you rely on "uninstalled" packages from an installation folder, as I said, it just confuses the concept of an installation folder.

Regarding your suggestion of using ⎕SE as a cache, that should already work, you ought to be able to use `LoadDependencies` with the target space being ⎕SE. Doing that would still require that you use aliases to resolve name conflicts, but feel free to try it out.

**Reply by DavinChurch** [reply]
I meant using LoadPackage for most everything instead of InstallPackage, so I didn't have an installation folder in the first place.

Thanks for the ⎕SE suggestion - I guess I hadn't put two and two together. I'll check it out.

**Reply by DavinChurch** [reply]
Ok, pre-loading packages in ⎕SE doesn't help much in my situation because I can't )SAVE the workspace if it contains references to ⎕SE. If I'm going to have to reconnect the packages every time I load then I might as well just put them in my WS every time. But that still means erasing them from the WS before every save (either to save space or disconnect ⎕SE links) and I SAVE frequently (even with SI where disconnections aren't really possible). So I'm still back to my original difficulties with too many WSes for me to manage.

Hmmm...
